File: scripts/test/wasm2c.py
# Copyright 2026 WebAssembly Community Group participants
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pathlib
import subprocess
import tempfile

from . import shared, support
from .shared import print_heading

spec_tests = [
]

def test_wasm2c_spec_output():
    for t in shared.options.spec_tests:
        test_path = pathlib.Path(t)
        if test_path.name not in spec_tests:
            continue

        print('..', test_path.name)

        is_fail_test = '.fail' in test_path.name

        test_subdir = f'wasm2c_spec_{test_path.stem}'
        test_subdir_path = pathlib.Path(test_subdir)
        test_subdir_path.mkdir(exist_ok=True)

        wasm2c_cmd = [shared.WASM2C[0], t, '-o', f'{test_subdir}/{test_path.stem}.c', '--allow-asserts']
        support.run_command(wasm2c_cmd, expected_status = (1 if is_fail_test else 0))

        c_sources = sorted(test_subdir_path.glob('*.c'))

        wasm_rt_dir = pathlib.Path(shared.options.binaryen_root) / 'src' / 'tools' / 'wasm2c' / 'wasm-rt'
        c_sources.append(wasm_rt_dir / 'wasm-rt-impl.c')
        c_sources.append(wasm_rt_dir / 'wasm-rt-mem-impl.c')
        c_sources.append(wasm_rt_dir / 'wasm-rt-exceptions-impl.c')

        compile_cmd = [shared.NATIVECC, '-O2', '-std=c11', '-D_GNU_SOURCE', '-D_DEFAULT_SOURCE', '-I.', f"-I{wasm_rt_dir}"] + [str(s) for s in c_sources] + ['-o', f'{test_subdir}/spec_test_runner']

        compile_cmd += ['-fno-optimize-sibling-calls', '-frounding-math']
        if 'gcc' in shared.NATIVECC.lower():
            compile_cmd.append('-fsignaling-nans')

        compile_cmd.append('-lm')
        compile_cmd.append('-lpthread')

        support.run_command(compile_cmd)

        # Run spec test runner binary and assert success
        support.run_command([f'{test_subdir}/spec_test_runner'])

def test_wasm2c_spec():
    print_heading('checking wasm2c spec testcases...')
    if shared.skip_if_on_windows('wasm2c-spec'):
        return
    test_wasm2c_spec_output()

File: src/tools/wasm2c/wasm2c.cpp
/*
 * Copyright 2026 WebAssembly Community Group participants
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//
// wasm2c console tool
//

#include "parser/wat-parser.h"
#include "parsing.h"
#include "support/colors.h"
#include "support/command-line.h"
#include "support/file.h"
#include "support/path.h"
#include "tools/tool-options.h"
#include "tools/wasm2c/assertion-emitter.h"
#include "tools/wasm2c/wasm2c-builder.h"
#include "wasm-io.h"

using namespace wasm;

struct Wasm2COptions : public ToolOptions {
  constexpr static const char* Wasm2COptionsCategory = "wasm2c options";

  std::string output;
  bool asserts = false;
  std::string prefix;
  std::string infile;

  Wasm2COptions()
    : ToolOptions(
        "wasm2c",
        "Transform .wasm/.wat files to standard C source and headers") {
    (*this)
      .add("--output",
           "-o",
           "Output file path (derives .h by changing extension, writes both to "
           "stdout if not specified)",
           Wasm2COptionsCategory,
           Arguments::One,
           [this](Options*, const std::string& argument) {
             output = argument;
             Colors::setEnabled(false);
           })
      .add("--allow-asserts",
           "",
           "Allow compilation of .wast testing asserts",
           Wasm2COptionsCategory,
           Arguments::Zero,
           [this](Options*, const std::string&) { asserts = true; })
      .add("--prefix",
           "-n",
           "Set the name prefix for the generated C symbols",
           Wasm2COptionsCategory,
           Arguments::One,
           [this](Options*, const std::string& argument) { prefix = argument; })
      .add_positional(
        "INFILE",
        Arguments::One,
        [this](Options*, const std::string& argument) { infile = argument; });
  }
};

int main(int argc, const char* argv[]) {
  Wasm2COptions options;
  options.parse(argc, argv);

  Wasm2CBuilder::Flags flags;
  flags.moduleName = options.prefix;

  if (options.asserts) {
    // When `--allow-asserts` is passed (i.e. when running spec tests), we need
    // to parse the extra s-expressions that come at the end of the `*.wast`
    // file after the module is defined. Therefore, in this case, we can't use
    // ModuleReader.

    auto input = read_file<std::string>(options.infile, Flags::Text);

    auto script = WATParser::parseScript(input);
    if (auto* err = script.getErr()) {
      Fatal() << err->msg;
    }

    Output cOut(options.output, Flags::Text);
    AssertionEmitter emitter(*script, flags, options.passOptions);
    emitter.emit(cOut.getStream(), options.output);
  } else {
    Module wasm;
    options.applyOptionsBeforeParse(wasm);
    try {
      ModuleReader reader;
      reader.read(options.infile, wasm, "");
    } catch (ParseException& p) {
      p.dump(std::cerr);
      Fatal() << "error in parsing input";
    }
    options.applyOptionsAfterParse(wasm);

    if (!options.output.empty()) {
      // Derive output header file path (.h instead of .c)
      std::string outputHPath = options.output;
      size_t lastDot = outputHPath.find_last_of('.');
      if (lastDot != std::string::npos) {
        outputHPath.replace(lastDot, std::string::npos, ".h");
      } else {
        outputHPath += ".h";
      }
      flags.headerName = Path::getBaseName(outputHPath);

      Output cOut(options.output, Flags::Text);
      Output hOut(outputHPath, Flags::Text);
      Wasm2CBuilder builder(flags);
      builder.processWasm(&wasm, cOut.getStream(), hOut.getStream());
    } else {
      // Write both to stdout sequentially
      std::cout << "/* === HEADER FILE === */\n";
      std::stringstream hStream;
      std::stringstream cStream;

      flags.headerName = "wasm.h"; // Default fallback
      Wasm2CBuilder builder(flags);
      builder.processWasm(&wasm, cStream, hStream);

      std::cout << hStream.str();
      std::cout << "\n/* === SOURCE FILE === */\n";
      std::cout << cStream.str();
    }
  }

  flush_and_quick_exit(0);
}
